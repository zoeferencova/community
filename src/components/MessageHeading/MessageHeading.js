import React, { Component } from "react";
import { PropTypes } from 'prop-types';
import { CHAT_DELETED } from "../../message-utils/events";
import ChatService from "../../services/chat-service";
import CommUnityContext from "../../contexts/context";
import MessageInfo from "../MessageInfo/MessageInfo";
import { ProfilePicture } from "../Utils/Utils";
import styles from "../MessageLayout/MessageLayout.module.css";

export default class MessageHeading extends Component {
    static contextType = CommUnityContext;

    state = {
        infoOpen: false,
        loading: false
    }
    
    // Displays a window popup confirming that chat is to be deleted
    // Deletes chat that is currently set to activeChat if confirmed
    // Removes chat in context and emits CHAT_DELETED event to socket which pushes chat deletion to the other member of the chat
    // Sets activeChat in context to null
    handleDelete = e => {
        e.preventDefault()

        this.setState({...this.state, loading: true })

        const chatId = this.context.activeChat.id;
        if (window.confirm('Are you sure you want to delete this chat? Your message history will be lost.')) {
            ChatService.deleteChat(chatId)
                .then(res => {
                    this.setState({...this.state, loading: false })
                    this.context.removeChat(chatId)
                    this.context.socket.emit(CHAT_DELETED, { chatId: chatId, receiverId: this.props.receiver.id })
                })
                .then(res => {
                    this.context.updateActiveChat(null)
                })
        }
    }

    // Sets activeChat in context to null which closes the currently opened chat
    handleClose(e) {
        e.preventDefault()
        this.context.updateActiveChat(null)
    }

    // Toggles the message info display
    toggleInfo = () => {
        this.state.infoOpen ? this.setState({ infoOpen: false }) : this.setState({ infoOpen: true })
    }

    render() {
        return (
            <div className={styles.headerContainer}>
                <div className={styles.chatHeader}>
                    <div className={styles.userInfo}>
                        <i className={`fas fa-chevron-left ${styles.chatBackArrow}`} onClick={this.props.mobileDisplayContacts}></i>
                        
                        <ProfilePicture className={styles.headerPic} first_name={this.props.receiver.first_name} />
                        <div className={styles.userName}>{this.props.receiver.first_name}</div>
                        <div className={styles.status}>
                            <div className={styles.indicator}></div>
                        </div>
                    </div>
                    <div className={styles.options}>
                        <i className="fas fa-info-circle" onClick={this.toggleInfo}></i>
                        <i className="fas fa-times" onClick={e => this.handleClose(e)}></i>
                    </div>
                </div>
                {this.state.infoOpen && <MessageInfo user={this.props.receiver} deleteChat={this.handleDelete} loading={this.state.loading} />}
            </div>
        );
    }
}

MessageHeading.propTypes = {
    receiver: PropTypes.object,
    mobileDisplayContacts: PropTypes.func
}