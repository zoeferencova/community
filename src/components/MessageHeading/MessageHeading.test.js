import React from "react";
import { render } from "enzyme";
import MessageHeading from "./MessageHeading";
import CommUnityContext from "../../contexts/context";
import mockContext from "../../contexts/mock-context";

describe("MessageHeading", () => {
    it("renders MessageHeading component given mock context data", () => {
        const mockProps = {
            receiver: {
                first_name: "James",
                id: 2
            }
        }
    
        const wrapper = render(
            <CommUnityContext.Provider value={mockContext}>
                <MessageHeading {...mockProps} />
            </CommUnityContext.Provider>
        );
        
        expect(wrapper).toMatchSnapshot();
    });
});