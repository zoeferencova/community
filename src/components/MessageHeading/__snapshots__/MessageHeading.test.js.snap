// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageHeading renders MessageHeading component given mock context data 1`] = `
<div
  class="headerContainer"
>
  <div
    class="chatHeader"
  >
    <div
      class="userInfo"
    >
      <i
        class="fas fa-chevron-left chatBackArrow"
      />
      <div
        class="circle indigo headerPic"
        first_name="James"
      >
        J
      </div>
      <div
        class="userName"
      >
        James
      </div>
      <div
        class="status"
      >
        <div
          class="indicator"
        />
      </div>
    </div>
    <div
      class="options"
    >
      <i
        class="fas fa-info-circle"
      />
      <i
        class="fas fa-times"
      />
    </div>
  </div>
</div>
`;
